beacon_node/eth1/src/cache.rs

```diff
@@ -0,0 +1,182 @@
+            }
+        }
+        Err(Error::InvalidData)
+    }
+
+    /// Returns a Vec<Eth1Data> corresponding to given distance range.
+    pub fn get_eth1_data_in_range(&self, start: u64, end: u64) -> Vec<Eth1Data> {
+        (start..end)
+            .map(|h| self.get_eth1_data(h))
+            .flatten() // Chuck Err values. This might be okay since its unlikely that the entire range returns None.
+            .collect::<Vec<Eth1Data>>()
+    }
+}
+
+fn fetch_eth1_data_in_range<F: Eth1DataFetcher>(
+    start: u64,
+    end: u64,
+    current_block_number: U256,
+    fetcher: Arc<F>,
+) -> impl Stream<Item = Result<(U256, Eth1Data)>, Error = Error> + Send {
+    stream::futures_ordered(
+        (start..end).map(move |i| fetch_eth1_data(i, current_block_number, fetcher.clone())),
+    )
+}
+
+/// Fetches Eth1 data from the Eth1Data fetcher object.
+fn fetch_eth1_data<F: Eth1DataFetcher>(
+    distance: u64,
+    current_block_number: U256,
+    fetcher: Arc<F>,
+) -> impl Future<Item = Result<(U256, Eth1Data)>, Error = Error> + Send {
+    let block_number: U256 = current_block_number
+        .checked_sub(distance.into())
+        .unwrap_or(U256::zero());
+    let deposit_root = fetcher.get_deposit_root(Some(BlockNumber::Number(block_number.as_u64())));
+    let deposit_count = fetcher.get_deposit_count(Some(BlockNumber::Number(block_number.as_u64())));
+    let block_hash = fetcher.get_block_hash_by_height(block_number.as_u64());
+    let eth1_data_future = deposit_root.join3(deposit_count, block_hash);
+    eth1_data_future.map(move |data| {
+        let eth1_data = Eth1Data {
+            deposit_root: data.0,
+            deposit_count: data.1?,
+            block_hash: data.2.ok_or(Error::DecodingError)?,
+        };
+        Ok((block_number, eth1_data))
+    })
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::web3_fetcher::Web3DataFetcher;
+    use std::time::{Duration, Instant};
+    use tokio::timer::{Delay, Interval};
+
+    // Note: Running tests using ganache-cli instance with config
+    // from https://github.com/ChainSafe/lodestar#starting-private-eth1-chain
+
+    fn setup() -> Web3DataFetcher {
+        let deposit_contract_address = "8c594691C0E592FFA21F153a16aE41db5beFcaaa";
+        let mut abi_path = std::env::current_dir().unwrap();
+        abi_path.push("deposit_contract.json");
+        let w3 = Web3DataFetcher::new("ws://localhost:8545", deposit_contract_address, abi_path);
+        return w3.unwrap();
+    }
+
+    #[test]
+    fn test_fetch() {
+        let w3 = setup();
+        let when = Instant::now() + Duration::from_millis(5000);
+        let task1 = Delay::new(when)
+            .and_then(|_| {
+                println!("Hello world!");
+                Ok(())
+            })
+            .map_err(|e| panic!("delay errored; err={:?}", e));
+        tokio::run(task1);
+        let task2 = fetch_eth1_data(0, 10.into(), Arc::new(w3)).and_then(|data| {
+            println!("{:?}", data);
```

Comment:
I think it's nice to avoid `println` in tests too.

Messages to describe the failure can be included in the `assert` calls. Other debugging print's can just be added on-demand.